components/script_bindings/str.rs

```diff
@@ -293,6 +298,62 @@ impl DOMString {
     }
 }
 
+/// Because this converts to a DOMString it become UTF-8 encoded which is closer to
+/// the spec definition of <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-json-bytes>
+/// but we generally do not operate on anything that is truly a WTF-16 string.
+///
+/// <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-a-json-string>
+pub fn serialize_jsval_to_json_utf8(
+    cx: SafeJSContext,
+    jsval: HandleValue,
+) -> Result<DOMString, Error> {
+    #[repr(C)]
+    struct VoidableString {
+        is_void: bool,
+        data: String,
+    }
+
+    let mut out_str = VoidableString {
+        is_void: true,
+        data: String::new(),
```

Comment:
Instead of two members here, I would just use `OnceCell<String>`
```suggestion
        string: OnceCell::new(),
```

Reply:
Ah, that's a lot cleaner and kinda what I had tried to do with an Option. Now that's thinking in Rust!